dataset/src/main/scala/frameless/SchemaWrapper.scala

```diff
@@ -0,0 +1,38 @@
+package frameless
+
+import shapeless.record.Record
+import shapeless.{HList, LabelledGeneric}
+
+/**
+  * Type-class to wrap non case class schema into Tuple1. For example: <br>
+  * Int                   -> Tuple1[Int] <br>
+  * List[Char]            -> Tuple1[Char] <br>
+  * case class X(i: Int)  -> No-wrapping <br>
+  * @tparam A the type to attempt to wrap
+  */
+trait SchemaWrapper[A] {
+  type N // The wrapped type
+  type Z <: HList
+}
+
+object SchemaWrapper extends LowPrioritySchemaWrapper{
+  type Aux[A, B, C] = SchemaWrapper[A]{ type N = B; type Z = C }
+
+  def apply[A](implicit schemaWrapper: SchemaWrapper[A]): SchemaWrapper[A] = schemaWrapper
+
+  // First attempt to derive a LabelledGeneric. If that fails it means this is not a case class based schema.
+  implicit def deriveSchemaWrapper[A, ARep <: HList]
+  (implicit
+   i0: LabelledGeneric.Aux[A, ARep]
```

Comment:
This looks rather expensive from a compile time perspective, maybe we could approximate that with `<: Product`?

Reply:
The Wrapper is used only inside the Exists type class. That type-class already derives LabelledGeneric. I make sure that the type is propagated so we don't have to extract it a second time. I actually spend more time trying to work around that than the original fix.